File: src/ads/demo/StatusDialog.cpp
//============================================================================
/// \file   StatusDialog.cpp
/// \author Uwe Kindler
/// \date   13.04.2020
/// \brief  Implementation of CStatusDialog class
//============================================================================

//============================================================================
//                                   INCLUDES
//============================================================================
#include "StatusDialog.h"

#include <iostream>

#include "DockManager.h"
#include "DockWidget.h"
#include "ui_StatusDialog.h"

/**
 * Private data class of CStatusDialog class (pimpl)
 */
struct StatusDialogPrivate
{
	CStatusDialog *_this;
	Ui::CStatusDialogClass ui;
	ads::CDockManager* DockManager;
	QMap<QString, ads::CDockWidget*> DockWidgets;

	/**
	 * Private data constructor
	 */
	StatusDialogPrivate(CStatusDialog *_public);
};
// struct StatusDialogPrivate

//============================================================================
StatusDialogPrivate::StatusDialogPrivate(CStatusDialog *_public) :
	_this(_public)
{

}

//============================================================================
CStatusDialog::CStatusDialog(ads::CDockManager* DockManager) :
	QDialog(DockManager),
	d(new StatusDialogPrivate(this))
{
	d->ui.setupUi(this);
	d->DockManager = DockManager;
	d->DockWidgets = DockManager->dockWidgetsMap();

	for (auto it = d->DockWidgets.begin(); it != d->DockWidgets.end(); ++it)
	{
		QVariant vDockWidget = QVariant::fromValue(it.value());
		d->ui.dockWidgetsComboBox->addItem(it.key(), vDockWidget);
	}
}

//============================================================================
CStatusDialog::~CStatusDialog()
{
	delete d;
}


//============================================================================
void CStatusDialog::on_dockWidgetsComboBox_currentIndexChanged(int index)
{
	if (index < 0)
	{
		return;
	}

	auto vDockWidget = d->ui.dockWidgetsComboBox->currentData();
	auto DockWidget = vDockWidget.value<ads::CDockWidget*>();
	d->ui.isClosedCheckBox->setChecked(DockWidget->isClosed());
	d->ui.isFloatingCheckBox->setChecked(DockWidget->isFloating());
	d->ui.tabbedCheckBox->setChecked(DockWidget->isTabbed());
	d->ui.isCurrentTabCheckBox->setChecked(DockWidget->isCurrentTab());
	d->ui.closableCheckBox->setChecked(DockWidget->features().testFlag(ads::CDockWidget::DockWidgetClosable));
	d->ui.movableCheckBox->setChecked(DockWidget->features().testFlag(ads::CDockWidget::DockWidgetMovable));
	d->ui.floatableCheckBox->setChecked(DockWidget->features().testFlag(ads::CDockWidget::DockWidgetFloatable));
	d->ui.deleteOnCloseCheckBox->setChecked(DockWidget->features().testFlag(ads::CDockWidget::DockWidgetDeleteOnClose));
	d->ui.customCloseHandlingCheckBox->setChecked(DockWidget->features().testFlag(ads::CDockWidget::CustomCloseHandling));
}

//---------------------------------------------------------------------------
// EOF StatusDialog.cpp

File: src/ads/demo/status_dialog.py
import os
import sys

from PyQt5 import uic
from PyQtAds import QtAds

UI_FILE = os.path.join(os.path.dirname(__file__), 'StatusDialog.ui')
StatusDialogUI, StatusDialogBase = uic.loadUiType(UI_FILE)

class CStatusDialog(StatusDialogUI, StatusDialogBase):
    
    def __init__(self, dock_manager: QtAds.CDockManager, parent=None):
        super().__init__(parent)

        self.setupUi(self)
        
        self.dock_manager = dock_manager
        self.dock_widgets = self.dock_manager.dockWidgetsMap()
        
        for key, widget in self.dock_widgets.items():
            self.dockWidgetsComboBox.addItem(key, widget)
            
    def on_dockWidgetsComboBox_currentIndexChanged(self, index: int):
        if not isinstance(index, int):
            return
        if index < 0:
            return
        
        dock_widget = self.dockWidgetsComboBox.currentData()
        self.isClosedCheckBox.setChecked(dock_widget.isClosed())
        self.isFloatingCheckBox.setChecked(dock_widget.isFloating())
        self.tabbedCheckBox.setChecked(dock_widget.isTabbed())
        self.isCurrentTabCheckBox.setChecked(dock_widget.isCurrentTab())
        self.closableCheckBox.setChecked(dock_widget.features() & QtAds.CDockWidget.DockWidgetClosable)
        self.movableCheckBox.setChecked(dock_widget.features() & QtAds.CDockWidget.DockWidgetMovable)
        self.floatableCheckBox.setChecked(dock_widget.features() & QtAds.CDockWidget.DockWidgetFloatable)
        self.deleteOnCloseCheckBox.setChecked(dock_widget.features() & QtAds.CDockWidget.DockWidgetDeleteOnClose)
        self.customCloseHandlingCheckBox.setChecked(dock_widget.features() & QtAds.CDockWidget.CustomCloseHandling)
